source/extensions/filters/network/redis_proxy/conn_pool_impl.cc

```diff
@@ -106,25 +111,73 @@ void InstanceImpl::ThreadLocalPool::onClusterRemoval(const std::string& cluster_
 
   // Treat cluster removal as a removal of all hosts. Close all connections and fail all pending
   // requests.
+  if (host_set_member_update_cb_handle_ != nullptr) {
+    host_set_member_update_cb_handle_->remove();
+    host_set_member_update_cb_handle_ = nullptr;
+  }
   while (!client_map_.empty()) {
     client_map_.begin()->second->redis_client_->close();
   }
+  while (!clients_to_drain_.empty()) {
+    (*clients_to_drain_.begin())->redis_client_->close();
+  }
 
   cluster_ = nullptr;
-  host_set_member_update_cb_handle_ = nullptr;
   host_address_map_.clear();
 }
 
+void InstanceImpl::ThreadLocalPool::onHostsAdded(
+    const std::vector<Upstream::HostSharedPtr>& hosts_added) {
+  for (const auto& host : hosts_added) {
+    std::string host_address = host->address()->asString();
+    // Insert new host into address map, possibly overwriting a previous host's entry.
+    host_address_map_[host_address] = host;
+    for (const auto& created_host : created_hosts_) {
+      if (created_host->address()->asString() == host_address) {
+        // Remove our "temporary" host created in makeRequestToHost().
+        onHostsRemoved({created_host});
+        created_hosts_.remove(created_host);
+        break;
+      }
+    }
+  }
+}
+
 void InstanceImpl::ThreadLocalPool::onHostsRemoved(
     const std::vector<Upstream::HostSharedPtr>& hosts_removed) {
   for (const auto& host : hosts_removed) {
     auto it = client_map_.find(host);
     if (it != client_map_.end()) {
-      // We don't currently support any type of draining for redis connections. If a host is gone,
-      // we just close the connection. This will fail any pending requests.
-      it->second->redis_client_->close();
+      if (it->second->redis_client_->active()) {
+        // Put the ThreadLocalActiveClient to the side to drain.
+        clients_to_drain_.push_back(std::move(it->second));
+        client_map_.erase(it);
+        if (!drain_timer_->enabled()) {
+          drain_timer_->enableTimer(parent_.config_.upstreamDrainPollIntervalInMs());
+        }
+      } else {
+        // There are no pending requests so close the connection.
+        it->second->redis_client_->close();
+      }
     }
-    host_address_map_.erase(host->address()->asString());
+    // There is the possibility that multiple hosts with the same address
+    // are registered in host_address_map_ given that hosts may be created
+    // upon redirection or supplied as part of the cluster's definition.
+    try {
+      if (host_address_map_.at(host->address()->asString()) == host) {
+        host_address_map_.erase(host->address()->asString());
+      }
+    } catch (std::out_of_range&) {
```

Comment:
would .find() be better here?  Since we don't know for sure that it's in the host_address_map_.

Reply:
there should be an entry with in host_address_map_ under the host's IP address and port string, but it may not point to the host object -- so at() should be as fast or faster than find() (they probably end up executing much the same code in this case). find() has an advantage of not throwing an exception...I'll change this code if you want.

Reply:
ok